trieste/acquisition/function/multi_objective.py

```diff
@@ -379,3 +384,245 @@ def _update_expected_improvement_fn(
                 isinstance(self._expected_improvement_fn, expected_hv_improvement), []
             )
             self._expected_improvement_fn.update(_partition_bounds)  # type: ignore
+
+
+class HippoPenalizationAcquisitionFunction(GreedyAcquisitionFunctionBuilder[ProbabilisticModel]):
```

Comment:
I expect you want to make this generic in the type of supported models. That way you can pass in a base builder that doesn't support all model types and still use this for those types of model.

```suggestion
class HippoPenalizationAcquisitionFunction(GreedyAcquisitionFunctionBuilder[ProbabilisticModelType]):
```

Reply:
thanks!

Reply:
I made this change, but the other types stayed as `ProbabilisticModel`, and I cannot fully claim understanding why it would't work with `ProbabilisticModelType` the whole way. Scanning other Trieste code didn't bring any clarity. Can you enlighten me please, where which type is supposed to go?

Reply:
Think of `ProbabilisticModelType` as meaning "a specific type of model", while `ProbabilisticModel` means "any model". Hence making HIPPO a `GreedyAcquisitionFunctionBuilder[ProbabilisticModelType]` is saying it might only support specific model types. Doing this only makes sense if you also change the base_acquisition_function_builder to be `AcquisitionFunctionBuilder[ProbabilisticModel]`. That allows you to pass in a base builder that only supports certain model types, as long as you only use HIPPO with those model types.

At the very least you'll want to change all mentions of `ProbabilsiticModel` in HIPPO (e.g. in prepare_acquisition_function) to `ProbabilsticModelType`. If you're getting typing errors feel free to message me and I can take a look.

Reply:
Upon further thinking, I think I get it, @uri-granta. This code would be valid when it should not be?
```
class HIPPO[ProbabilisticModel]:
    self.base_builder: AcquisitionBuilder[ProbabilisticModel]
    
    def __init__(self, base_builder):
        self.base_builder = base_builder

ehvi = EHVI[B]()
hippo = HIPPO[A](envi)
```

whereas if we were to declare HIPPO like that
```
class HIPPO[ProbabilisticModelType]
```

the above code would indeed be caught by mypy as invalid, right? Because the types of base builder and hippo don't match, and `ProbabilisticModelType` controls for it.

Is that understanding correct? If so, I certainly agree we shall use `ProbabilisticModelType` to ensure penalization and base functions are typed the same way

Reply:
If we don't use `ProbabilisticModelType` then mypy should complain if we try to pass in a base builder like `AugmentedExpectedImprovement` or `PredictiveVariance`, since those builders don't support all `ProbabilisticModel`s.

Would you like me to have a go at updating this so you can see what I mean?

Reply:
Got it!

We are only interested in multi-objective non-batch builders, which support any models at the moment (EHVI and ECHVI). But clearly we don't want to be restrictive in case later on we have some multi-objective builders that only support a subset of models. I guess it's worth to use `ProbabilisticModelType` here.

Thanks for the offer. I'll give it another go first, watch out for the update (likely tomorrow).

Reply:
updated